docs/dev_plan.md

```diff
@@ -0,0 +1,339 @@
+design should include data structures and API's that are general enough
+to support integration of third party libraries as well as the ability
+to change any library for another that accomplishes a similar task
+even if by alternative methods.
+
+### Performance first
+
+A key design consideration of OpenTurbine is computational efficiency
+or performance. Both the quantity of work and the efficiency of data should
+be considered and measured (i.e., profiled) during any development effort.
+A modular architecture should support offloading computationally expensive
+tasks to hardware accelerators or specialized libraries, and support
+multiple options for doing so depending on user configurations and
+the computational environment.
+
+Similar to modularity in the architecture, expensive tasks should be
+structured in a kernel form. This low-level design pattern combines
+expensive mathematical operations into an aggregate form, and structures
+them so that performance libraries or compilers can parallelize the
+computation. This pattern helps to encapsulate expensive operations and
+algorithms. Additionally, it follows the modular architecture design
+described above in that it supports swapping accelerators or parallelization
+methods../
+
+
+#### Data-oriented design
+
+OpenTurbine developers designing new algorithms and data structures
+should become familiar with the concepts of [data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design),
+particularly [structures of arrays vs. arrays of structures](https://stackoverflow.com/questions/17924705/structure-of-arrays-vs-array-of-structures).
+The key concept of this paradigm is to structure data so that it maps
+closely to the form it will be represented and used within the relevant
+processing units. For example, it is a common mistake in object-oriented
+design to represent data as arrays of structures (AoS),
+and this can lead to inefficiency in CPU cache
+usage since aggregate operations such as matrix math must load
+data from non-contiguous memory rather than perform the operations
+in a vector form. In a data-oriented design, data is more
+commonly represented as structures of arrays (SoA) where contiguous
+portions of memory are loaded into the cache and operated on
+either by the CPU or a hardware accelerator. The graphic below
+illustrates the difference between structures of arrays and arrays
+of structures.
+
+```{image} images/AoS_SoA.pdf
+:alt: aos_soa
+:width: 400px
+:align: center
+```
+
+The difference in usage for the above two patterns is illustrated via
+an example Python-based math operation below where a series of points
+are shifted in one direction by a constant amount.
+
+```python
+class Point:
+    x: int
+    y: int
+    z: int
+
+# Initialize test data
+all_points = [Point(random(i), random(i), random(i)) for i in range(100)]
+
+# Array of Structures
+AOS = [all_points[i].y for i in range(100)]
+
+# Structure of Arrays
+SOA = all_points.y + 10
+```
```

Comment:
Are we going to develop in python? Perhaps this example should be illustrated in whichever language we will use. Python array syntax is not very clear to anyone who doesn't already know python.

Reply:
It's a good point. I've removed this (see above comment), and we can add more meaningful examples in another section, if needed.